signer/src/storage/postgres.rs

```diff
@@ -92,8 +97,136 @@ impl TryFrom<&NakamotoBlock> for model::StacksBlock {
 
 impl PgStore {
     /// Connect to the Postgres database at `url`.
-    pub async fn connect(url: &str) -> Result<Self, sqlx::Error> {
-        Ok(Self(sqlx::PgPool::connect(url).await?))
+    pub async fn connect(url: &str) -> Result<Self, Error> {
+        Ok(Self(
+            sqlx::PgPool::connect(url)
+                .await
+                .map_err(Error::SqlxConnect)?,
+        ))
+    }
+
+    /// Apply the migrations to the database.
+    // Note 1: This could be generalized and moved up to the `storage` module, but
+    // left that for a future exercise if we need to support other databases.
+    // Note 2: The `sqlx` "migration" feature results in dependency conflicts
+    // with sqlite from the clarity crate.
+    // Note 3: The migration code paths have no explicit integration tests, but are
+    // implicitly tested by all integration tests using `new_test_database()`.
+    pub async fn apply_migrations(&self) -> Result<(), Error> {
+        tracing::info!("Preparing to run database migrations");
+
+        sqlx::raw_sql(
+            r#"
+                CREATE TABLE IF NOT EXISTS public.__sbtc_migrations (
+                    key TEXT PRIMARY KEY
+                );
+            "#,
+        )
+        .execute(&self.0)
+        .await
+        .map_err(Error::SqlxMigrate)?;
+
+        let mut trx = self
+            .pool()
+            .begin()
+            .await
+            .map_err(Error::SqlxBeginTransaction)?;
+
+        // Collect all migration scripts and sort them by filename. It is important
+        // that the migration scripts are named in a way that they are executed in
+        // the correct order, i.e. the current naming of `0001__`, `0002__`, etc.
+        let mut migrations = PGSQL_MIGRATIONS.files().collect::<Vec<_>>();
+        migrations.sort_by_key(|file| file.path().file_name());
+        for migration in migrations {
```

Comment:
Have you thought about using something like [refinery](https://github.com/rust-db/refinery) for this? Although what is here seems to work, it seems like we can outsource much of this to that crate. And in the event of any migration issues, we'd probably want to add a few more columns to the migrations table to make debugging easier. So maybe it's better to inherit those features from an external crate. Thoughts?

Reply:
I wasn't aware of `refinery` and didn't see it when I did a quick search, but it looks nice! Sad it doesn't natively support sqlx and you have to do some workarounds. I'm all-for using an external crate otherwise (I briefly considered `diesel_migrations` but decided against it at this stage to avoid bringing in additional deps since these migrations are for now quite simple).

Idunno, for me "right now" I just felt like we needed some migration logic, and this "works for now" (though it maybe isn't the most advanced). I'd be interested in tabling it for now if this works in general and looking at "upgrading" when we start nearing a public release? Thoughts on that?

Reply:
That works. Let's do that.

Reply:
Added #537 to revisit